log.go

```diff
@@ -20,9 +20,38 @@
 
 package fx
 
-import "go.uber.org/fx/internal/fxlog"
+import (
+	"errors"
 
-// WithLogger exposes logger option for tests.
-func WithLogger(l fxlog.Logger) Option {
-	return withLogger(l)
+	"go.uber.org/fx/internal/fxlog"
+)
+
+// logBuffer will buffer all messages until a logger has been
+// initialized.
+type logBuffer struct {
+	entries []fxlog.Entry
+	logger  fxlog.Logger
+}
+
+// Log buffers or logs an entry.
+func (l *logBuffer) Log(entry fxlog.Entry) {
+	if l.logger == nil {
+		l.entries = append(l.entries, entry)
+	} else {
+		l.logger.Log(entry)
+	}
+}
+
+// Connect flushes out all buffered entries to a logger and resets them.
+func (l *logBuffer) Connect(logger fxlog.Logger) error {
+	if logger == nil {
+		return errors.New("missing a logger to connect to")
+	}
+	l.logger = logger
+	for _, e := range l.entries {
+		logger.Log(e)
+	}
+	l.entries = []fxlog.Entry{}
```

Comment:
nil is a valid slice and costs no allocs

```suggestion
	l.entries = nil
```